pooch/utils.py

```diff
@@ -210,3 +212,58 @@ def parse_url(url):
     parsed_url = urlsplit(url)
     protocol = parsed_url.scheme or "file"
     return {"protocol": protocol, "netloc": parsed_url.netloc, "path": parsed_url.path}
+
+
+def make_local_storage(path, env=None, version=None):
+    """
+    Create the local cache directory and make sure it's writable.
+
+    If the directory doesn't exist, it will be created.
+
+    Parameters
+    ----------
+    path : str, PathLike, list or tuple
+        The path to the local data storage folder. If this is a list or tuple,
+        we'll join the parts with the appropriate separator. Use
+        :func:`pooch.os_cache` for a sensible default.
+    version : str or None
+        The version string for your project. Will be appended to given path if
+        not None.
+    env : str or None
+        An environment variable that can be used to overwrite *path*. This
+        allows users to control where they want the data to be stored. We'll
+        append *version* to the end of this value as well.
+
+    Returns
+    -------
+    local_path : PathLike
+        The path to the local directory.
+
+    """
+    if env is not None and env in os.environ and os.environ[env]:
+        path = os.environ[env]
+    if isinstance(path, (list, tuple)):
+        path = os.path.join(*path)
+    if version is not None:
+        path = os.path.join(str(path), version)
+    path = os.path.expanduser(str(path))
+    # Check that the data directory is writable
+    try:
+        if not os.path.exists(path):
+            action = "create"
+            os.makedirs(path)
+        else:
+            action = "write to"
+            tempfile.NamedTemporaryFile(dir=path)
+    except PermissionError:
+        message = (
+            "Cannot %s data cache folder '%s'. "
+            "Will not be able to download remote data files. "
+        )
+        args = [action, path]
+        if env is not None:
+            message += "Use environment variable '%s' to specify another directory."
+            args += [env]
+
+        get_logger().warning(message, *args)
```

Comment:
I realize that I'm the one who made this a warning when we switched over to using the logging API, but do you think it should become an error? If the cache directory isn't writable, the user will just hit another error right after when they try to download something. But the real cause is here and passing through with only a warning could just obscure that.

Reply:
I was thinking the same thing. If this is the case, the warning will pop up on `import` (usually when a package creates the `Pooch` instance) and an exception is only raised on `Pooch.fetch`. So if running on Jupyter/IPython there will be time to act and set the environment variable. But if running on a script then the error comes right after the warning and people might not see it.

In that case, it might be better to added the message to the error and re-raise it. What do you think?

Reply:
I was only thinking about the case of running it in a script and hadn't thought about jupyter. In any case I think that changing it from warning to error and re-raising the exception is the way to go.

Reply:
Agreed. I opened #150 to track this for a future PR instead of implementing it here to avoid changing 2 things at once. Thanks for the feedback @danshapero :+1: